### PR TITLE
:sparkles: low-code: Add Incremental Parent State Handling to SubstreamPartitionRouter

### PR DESCRIPTION
**Pull request by tolik0**
## What
This PR resolves the issue with incremental substream reads in the low-code. It ensures that the state is passed to the parent stream when generating partitions to avoid reading the parent stream in full refresh mode. Resolves: https://github.com/airbytehq/airbyte-internal-issues/issues/7369

Connector's changes:
- [Jira](https://github.com/airbytehq/airbyte/pull/38612)
- [Slack](https://github.com/airbytehq/airbyte/pull/38618)
- [Zendesk Support](https://github.com/airbytehq/airbyte/pull/38653)

## How
The code changes introduce a new field `incremental_dependency` in the `ParentStreamConfig` class. The `SubstreamPartitionRouter` was updated to read parent state incrementally and save the parent state when the stream slice is processed. Additionally, new default methods were added to StreamSlicer for cases when a stream slicer does not have parent streams.

## User Impact

- The end result is an optimized incremental substream reading process, which improves sync performance by reducing the need to read the parent stream in full refresh mode.
- There are no negative side effects anticipated.

## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [x] NO ❌

### CODE REVIEW

**Comment by tolik0** [comment]
@artem1205 I will create separate PRs for connectors; I just added these changes here for now for easier testing. This way we can execute regression tests from this branch.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py:32]: this method is only used on partition routers so I think it would be worth creating a new `PartitionRouter` interface that would be more specific.

We should also leave the two methods as abstract to ensure developers explicitly define them instead of relying on a default behavior that may or may not make sense for their use case

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py:34]: the docstring should be more specific about the parameter expectations. stream_state must have `parent_state` which is not obvious from the signature. It might even be worth creating a new class to codify the expectation (kind of like how StreamSlice codifies the presence of a partition and a cursor_slice)

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py:167]: why is `stream_state` an `Optional`?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:89]: this deserves a comment as it's not obvious why we need to update the router's parent state

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py:32]: Even though this method is mostly used by partition routers, it's not the only use case. The CartesianProductStreamSlicer can use it too. Check it out here: [CartesianProductStreamSlicer Example](https://github.com/airbytehq/airbyte/blob/5172a3247eb3ed082d1c1adb56add6922301dff7/airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py#L655).

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py:34]: Updated docs for all classes

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py]: Fixed

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:119]: Added comment

Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:119]: Stream slicers and partition routers don't keep track of the cursor and don't have a `set_initial_state` method. `set_parent_state` sets the state to the parent streams that partition routers operate on, if such parent streams exist.

Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:119]: seems to me this means we want to add a `set_initial_state` method to the routers - from the caller's perspective, it doesn't really matter that it's setting a parent state. It's an implementation detail. 

Furthermore, not all stream slicers know or care about parents (eg list partition router)

Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:119]: edit: I do think the state being set is somewhat different, but the interface feels very strange to me for two reasons
1. Not every stream slicers should know or care about parent streams
2. even from the caller, the fact that it's updating a parent stream state is irrelevant. [It's just passing the whole state object](https://github.com/airbytehq/airbyte/pull/38211/files#diff-cb24ede0b6b09974da9a16d25d8fa5f772ae497c742224f08dac1bdc1a903f9dR119)

Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:119]: The same goes with `set_initial_state` for stream slicers, as not every stream slicer cares about the state. Introducing a `set_initial_state` method could create confusion, implying that partition routers are somehow dependent on the state. `set_parent_state` clearly articulates that it is only connected to the parent streams, as the stream interacts with the parent stream through the partition router.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py:32]: `DeclarativeCursor` doesn't, right?

I think it makes sense for `CartesianProductStreamSlicer` for also being a `PartitionRouter`

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py:32]: let's leave the method as abstract to force the concrete class to implement it (or intentionally implement a no-op)

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py:53]: let's leave the method as abstract to force the concrete class to implement it (or intentionally implement a no-op)

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:119]: yeah, that's exactly the point - we're now saying that `SubstreamPartitionRouter` and `CartesianProductStreamSlicer` are incremental and have a concept of state they need to keep track of.

I'm not sure if it'll make the code simpler or more complicated, but conceptually, this makes me think the `SubstreamPartitionRouter` should have a `Cursor` when `incremental_dependency` is True.

This would essentially mean that instead of keeping the parent state in a dict, it could pass the parent records to its cursor's observe method while iterating over them. 

I feel 4/10 strongly on this issue so not worth blocking on.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py:141]: let's add a comment here explaining that the sync mode and the stream state passed are irrelevant since we set the initial state from `set_parent_state`

**Review by artem1205 — DISMISSED** [review]
LGTM! could you please add more tests to ensure that changes and stream state is handled correctly.
e.g. `test_incremental_parent_state`

1. create test_source
2. create stream_A (parent)
3. create stream_B (child)
4. run test_source.read() with/without stream_state
5. assert: 
    - requests
    - records
    - stream_state (in output)

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py]: Changed CartesianProductStreamSlicer to PartitionRouter

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py]: I changed the methods' names. Regarding using a Cursor to keep track of the state, it is already managed by the cursor of parent streams. Partition routers just retrieve the state from the internal parent streams.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py]: Fixed

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/stream_slicer.py]: Fixed

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py:143]: Fixed

**Review by brianjlai — COMMENTED** [review]
I have some questions around how files were programmatically generated from the changes to the schema.

Also, is the plan to also update the CDK for the connectors being update? Should they be split out into a follow up PR?

File: airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/manifest.yaml
Comment (tolik0) [reply airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/manifest.yaml:1]: No, I accidentally pushed this line. It is a fix for the local issue.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:110]: Fixed

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/models/declarative_component_schema.py
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/models/declarative_component_schema.py]: Fixed

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/cartesian_product_stream_slicer.py
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/cartesian_product_stream_slicer.py]: Fixed

**Comment by tolik0** [comment]
@brianjlai, I have linked the connectors' PRs to the description. I added these changes to this PR so it can be easily tested using the regression tool from this branch. I will delete these changes before merging. However, regarding the connectors PR, it would be better to use a global parent state for them as it will help to avoid breaking changes and concerns about the new state size.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml:2007]: I'm trying to think if there's a more descriptive word for this field. `incremental_dependency` doesn't provide a clear meaning to what this is doing.

Maybe something like `incremental_sync_based_on_child`? I'm not attached to it and the naming is not a blocker, but trying to consider how we can make it more clear what this boolean does.

**Review by brianjlai — APPROVED** [review]
Looks good, and makes sense for convenience that these are in the PR for testing and validation purposes. Do you have any live test runs against the connectors you are implementing this for and using `--use-local-cdk`?

Once that is validated I think this is ready to be merged, did you address @artem1205 's changes. It looked like it, but I noticed he still had it marked as changes requested

**Comment by natikgadzhi** [comment]
@tolik0, let me know when you separate out connector changes and rebase — I'm happy to get this merged this week if @artem1205 is still away, I have the gawd-tier merge button privileges.

**Comment by tolik0** [comment]
@natikgadzhi @brianjlai I've attached an example of the regression test results. The target version shows more records due to the addition of lookback windows, yet it reproduces all records from the control version.

Given that state compatibility affects the regression tests for the connectors in this PR, the best way to validate these changes is through manual verification. Specifically, ensure the parent stream is using the correct cursor, as demonstrated in the new test I've added.

Regarding Artem's request, he only asked for the test that I have already added.

Screenshot from 2024-06-13 13-20-18
Screenshot from 2024-06-13 13-20-35

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py]: Sorry, I missed that. Fixed

**Pull request merged**
